pkg/csi/service/node.go

```diff
@@ -680,109 +690,272 @@ func (driver *vsphereCSIDriver) NodeGetInfo(
+		csiNodeTopologyInstance, ok := event.Object.(*csinodetopologyv1alpha1.CSINodeTopology)
+		if !ok {
+			log.Debugf("Received unidentified object - %+v", event.Object)
+			continue
 		}
-		defer func() {
-			if vcManager != nil {
-				err = vcManager.UnregisterAllVirtualCenters(ctx)
-				if err != nil {
-					log.Errorf("UnregisterAllVirtualCenters failed. err: %v", err)
+		switch csiNodeTopologyInstance.Status.Status {
+		case csinodetopologyv1alpha1.CSINodeTopologySuccess:
+			accessibleTopology := make(map[string]string)
+			topologyLabels := csiNodeTopologyInstance.Status.TopologyLabels
+			if len(topologyLabels) > 0 {
+				for _, label := range topologyLabels {
+					accessibleTopology[label.Key] = label.Value
 				}
 			}
-		}()
-		//Connect to vCenter
-		err = vcenter.Connect(ctx)
-		if err != nil {
+			return accessibleTopology, nil
+		case csinodetopologyv1alpha1.CSINodeTopologyError:
 			return nil, logger.LogNewErrorCodef(log, codes.Internal,
-				"failed to connect to vcenter host: %s. err: %v", vcenter.Config.Host, err)
+				"failed to retrieve topology information for Node: %q. Error: %q", nodeID,
+				csiNodeTopologyInstance.Status.ErrorMessage)
 		}
-		// Get VM UUID
-		uuid, err := getSystemUUID(ctx)
-		if err != nil {
-			return nil, logger.LogNewErrorCodef(log, codes.Internal,
-				"failed to get system uuid for node VM. err: %v", err)
+	}
+	return nil, logger.LogNewErrorCodef(log, codes.Internal,
+		"timed out while waiting for topology labels in CSINodeTopology instance with name %q to be updated.", nodeID)
+}
+
+func getNodeObject(ctx context.Context, nodeName string) (*v1.Node, error) {
+	log := logger.GetLogger(ctx)
+
+	// Create the kubernetes client.
+	k8sClient, err := k8s.NewClient(ctx)
+	if err != nil {
+		log.Errorf("failed to create K8s client. Error: %v", err)
+		return nil, err
+	}
+	// Fetch node object.
+	nodeObj, err := k8sClient.CoreV1().Nodes().Get(ctx, nodeName, metav1.GetOptions{})
+	if err != nil {
+		log.Errorf("failed to fetch node object with name %q. Error: %v", nodeName, err)
+		return nil, err
+	}
+	return nodeObj, nil
+}
+
+// getCSINodeTopologyWatchTimeoutInMin returns the timeout for watching
+// on CSINodeTopology instances for any updates.
+// If environment variable NODEGETINFO_WATCH_TIMEOUT_MINUTES is set and valid,
+// return the interval value read from environment variable
+// otherwise, use the default timeout 1 min.
+func getCSINodeTopologyWatchTimeoutInMin(ctx context.Context) int {
+	log := logger.GetLogger(ctx)
+	watcherTimeoutInMin := defaultTopologyCRWatcherTimeoutInMin
+	if v := os.Getenv("NODEGETINFO_WATCH_TIMEOUT_MINUTES"); v != "" {
+		if value, err := strconv.Atoi(v); err == nil {
+			if value <= 0 {
+				log.Warnf("Timeout set in env variable NODEGETINFO_WATCH_TIMEOUT_MINUTES %s is equal or "+
+					"less than 0, will use the default timeout of %d minute(s)", v, watcherTimeoutInMin)
+			} else {
+				watcherTimeoutInMin = value
```

Comment:
Can we add a check for max allowed value?

Reply:
Done